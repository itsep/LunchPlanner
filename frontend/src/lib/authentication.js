import Cookie from 'js-cookie'
import { withLunchspaceSubdomain } from './lunchspace_subdomain'
import apiFetch from './api_fetch'
import routeLocations from '../pages/route_locations'


const isProbablyAuthenticatedCookieName = 'authenticated'
/**
 * checks if a token exists as a cookie
 */
export function isProbablyAuthenticated() {
  return Cookie.get(isProbablyAuthenticatedCookieName) === '1'
}
/**
 * checks if a token does not exists as a cookie
 */
export function isDefinitelyNotAuthenticated() {
  return Cookie.get(isProbablyAuthenticatedCookieName) !== '1'
}

export function logout() {
  return apiFetch('/api/account/logout', {
    method: 'POST',
  }).then(() => {
    window.location = withLunchspaceSubdomain(routeLocations.LOGIN)
  })
}
