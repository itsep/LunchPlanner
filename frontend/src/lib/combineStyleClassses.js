export default function (...rest) {
  return rest.join(' ')
}
