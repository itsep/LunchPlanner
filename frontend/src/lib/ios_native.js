export const headerBarClassName = 'app-header-bar'
