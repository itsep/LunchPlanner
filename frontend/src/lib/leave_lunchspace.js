import apiFetch from './api_fetch'
import routeLocations, { redirect } from '../pages/route_locations'

export function redirectToLunchspaces() {
  redirect(routeLocations.LUNCHSPACES)
}

export function leaveLunchspace(redirectCallback) {
  return () => apiFetch('/api/lunchspace/leave', {
    method: 'DELETE',
  }).then(redirectCallback, redirectCallback)
}

export const leave = leaveLunchspace(redirectToLunchspaces)
