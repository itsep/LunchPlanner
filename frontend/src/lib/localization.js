import LocalizedStrings from 'react-localization'

import { languageKeyToLocalisationMap } from 'shared/localizations/index'

const localizedStrings = new LocalizedStrings(languageKeyToLocalisationMap)

export default localizedStrings
