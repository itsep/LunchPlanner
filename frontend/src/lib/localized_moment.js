import moment from 'moment'
import 'moment/locale/de'
import localizedStrings from './localization'

moment.updateLocale('en', {
  calendar: {
    lastDay: '[Yesterday]',
    sameDay: '[Today]',
    nextDay: '[Tomorrow]',
    lastWeek: '[last] dddd',
    nextWeek: 'dddd',
    sameElse: 'L',
  },
})
moment.updateLocale('de', {
  calendar: {
    lastDay: '[gestern]',
    sameDay: '[heute]',
    nextDay: '[morgen]',
    lastWeek: '[letzten] dddd',
    nextWeek: 'dddd',
    sameElse: 'L',
  },
})

moment.locale(localizedStrings.getLanguage())

export default moment
