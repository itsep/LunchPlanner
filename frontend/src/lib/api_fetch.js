import { toLocalizableError } from 'shared/lib/error'
import { withLunchspaceSubdomain } from './lunchspace_subdomain'
import { redirectToLogin } from './authentication'
import { catchAuthenticationErrorAndRedirect, catchAuthorizationErrorAndRedirect } from './auto_redirect'

const reject = Promise.reject.bind(Promise)
/**
 *
 * @param url
 * @param config {Object}
 * @returns {Promise<{data: any, response: Response}, LocalizableError>}
 */
export default function apiFetch(url, config = {}) {
  const redirectToLoginCallback = config.redirectToLogin || redirectToLogin
  return Promise.resolve().then(() => {
    const init = Object.assign({}, {
      credentials: 'same-origin',
    }, config)
    init.headers = Object.assign({}, {
      'content-type': 'application/json',
    }, config.headers)

    if (config.body) {
      init.body = JSON.stringify(config.body)
    }
    return fetch(withLunchspaceSubdomain(url), init).then((response) => {
      // everything okay
      if (response.ok) {
        return response.json().then(data => ({ data, response }))
      }
      // an error occured
      return response.json().then(reject)
    })
  })
    // convert every error to an localizable error
    .catch(errorObject => reject(toLocalizableError(errorObject)))
    // catch authentication errors and redirect to login if needed
    .catch(catchAuthenticationErrorAndRedirect.bind(null, redirectToLoginCallback))
    .catch(catchAuthorizationErrorAndRedirect.bind(null, redirectToLoginCallback))
}
