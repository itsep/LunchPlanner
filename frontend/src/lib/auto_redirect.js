import { AuthenticationError, AuthorizationError } from 'shared/lib/error'

const makeNeverResolvingPromise = () => new Promise(() => {})

export function catchAuthenticationErrorAndRedirect(redirect, error) {
  if (error instanceof AuthenticationError) {
    redirect()
    return makeNeverResolvingPromise()
  }
  throw error
}

export function catchAuthorizationErrorAndRedirect(redirect, error) {
  if (error instanceof AuthorizationError) {
    redirect()
    return makeNeverResolvingPromise()
  }
  throw error
}
