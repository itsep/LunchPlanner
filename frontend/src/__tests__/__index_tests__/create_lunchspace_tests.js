import React from 'react'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'
import { escapeSubdomain } from 'shared/lib/subdomain'
import { LocalizableError } from 'shared/lib/error/localizable_error'
import CreateLunchspace from '../../pages/create_lunchspace/components/create_lunchspace'


it('Create Lunchspace renders correctly', () => {
  // fetch.once(JSON.stringify({ count: 0 }))
  const tree = renderer
    .create(<CreateLunchspace />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

describe('create lunchspace', () => {
  const lunchspaceName = 'VSF Experts Mannheim'
  const lunchspaceSubdomain = escapeSubdomain(lunchspaceName)
  const shortSubdomainName = 'vsf-experts-ma'
  let createLunchspace
  let nameInput
  beforeEach(() => {
    createLunchspace = mount(shallow(<CreateLunchspace />).get(0))
    nameInput = createLunchspace.find("input[name='lunchspace-name']")
    nameInput.instance().value = lunchspaceName
    nameInput.simulate('change')
  })
  it('update subdomain correctly', async () => {
    expect(createLunchspace.state('lunchspaceName')).toEqual(lunchspaceName)
    expect(createLunchspace.state('lunchspaceSubdomain')).toEqual(lunchspaceSubdomain)
  })
  it('update subdomain without chaining the name', async () => {
    const subdomainInput = createLunchspace.find("input[name='subdomain']")
    subdomainInput.instance().value = shortSubdomainName
    subdomainInput.simulate('change')

    expect(createLunchspace.state('lunchspaceName')).toEqual(lunchspaceName)
    expect(createLunchspace.state('lunchspaceSubdomain')).toEqual(shortSubdomainName)
  })
  it('update name without chaining the subdomain', async () => {
    // change subdomain to a custom name
    const subdomainInput = createLunchspace.find("input[name='subdomain']")
    subdomainInput.instance().value = shortSubdomainName
    subdomainInput.simulate('change')
    expect(createLunchspace.state('lunchspaceName')).toEqual(lunchspaceName)
    expect(createLunchspace.state('lunchspaceSubdomain')).toEqual(shortSubdomainName)
    // remove the last char of the name
    const nameWithoutLastChar = lunchspaceName.substring(0, lunchspaceName.length - 1)
    nameInput.instance().value = nameWithoutLastChar
    nameInput.simulate('change')
    // subdomain should not change
    expect(createLunchspace.state('lunchspaceName')).toEqual(nameWithoutLastChar)
    expect(createLunchspace.state('lunchspaceSubdomain')).toEqual(shortSubdomainName)
  })
  it('create a new lunchspace successful', async () => {
    fetch.mockResponse(JSON.stringify({ lunchspaceName, lunchspaceSubdomain }))
    await createLunchspace.instance().handleSubmit()

    expect(createLunchspace.state('isLoading')).toBeFalsy()
    expect(createLunchspace.state('lunchspaceCreated')).toBeTruthy()
  })
  it('create new lunchspace failure - lunchspace subdomain already exists', async () => {
    const errorMessage = 'Lunchspace subdomain already exists.'
    fetch.mockResponse(JSON.stringify({ message: errorMessage }), { status: 409 })
    await createLunchspace.instance().handleSubmit()

    expect(createLunchspace.state('isLoading')).toBeFalsy()
    expect(createLunchspace.state('lunchspaceCreated')).toBeFalsy()
    expect(createLunchspace.state('error')).toBeDefined()
    expect(createLunchspace.state('error')).toBeInstanceOf(LocalizableError)
    expect(createLunchspace.state('lastError')).toBeDefined()
    expect(createLunchspace.state('lastError')).toEqual(createLunchspace.state('error'))

  })
})

