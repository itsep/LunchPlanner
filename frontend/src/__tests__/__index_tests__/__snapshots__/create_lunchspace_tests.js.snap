// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Lunchspace renders correctly 1`] = `
<div
  className="FormSection-root-3"
>
  <div
    className="MuiCollapse-container-4 MuiCollapse-entered-5"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper-6"
    >
      <div
        className="MuiCollapse-wrapperInner-7"
      >
        <form
          onSubmit={[Function]}
        >
          <h2
            className="MuiTypography-root-8 MuiTypography-title-14"
          >
            Create Lunchspace
          </h2>
          <div
            aria-describedby={undefined}
            className="MuiFormControl-root-32 MuiFormControl-marginNormal-33 CreateLunchspace-textField-1"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <label
              className="MuiFormLabel-root-41 MuiInputLabel-root-36 MuiInputLabel-formControl-37 MuiInputLabel-animated-40"
              data-shrink={false}
              htmlFor={undefined}
            >
              Lunchspace Name
            </label>
            <div
              className="MuiInput-root-46 MuiInput-formControl-47 MuiInput-underline-50"
            >
              <input
                aria-invalid={false}
                aria-required={false}
                autoComplete="organization"
                autoFocus={true}
                className="MuiInput-input-54"
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                name="lunchspace-name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={undefined}
                onKeyUp={undefined}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                rows={undefined}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="MuiFormControl-root-32 CreateLunchspace-textField-1"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <label
              className="MuiFormLabel-root-41 MuiInputLabel-root-36 MuiInputLabel-formControl-37 MuiInputLabel-animated-40 MuiInputLabel-shrink-39"
              data-shrink={true}
              htmlFor="input-with-icon-adornment"
            >
              Subdomain to your Lunchspace
            </label>
            <div
              className="MuiInput-root-46 MuiInput-formControl-47 MuiInput-underline-50"
            >
              <input
                aria-invalid={false}
                aria-required={false}
                autoComplete="off"
                autoFocus={undefined}
                className="MuiInput-input-54"
                defaultValue={undefined}
                disabled={false}
                id={undefined}
                name="subdomain"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={undefined}
                onKeyUp={undefined}
                placeholder="your-space-url"
                readOnly={undefined}
                required={undefined}
                rows={undefined}
                type="text"
                value=""
              />
              <div
                className="MuiInputAdornment-root-59 MuiInputAdornment-positionEnd-61"
              >
                <p
                  className="MuiTypography-root-8 MuiTypography-body1-17 MuiTypography-colorTextSecondary-30"
                >
                  .lunchspace.de
                </p>
              </div>
            </div>
            <p
              className="MuiFormHelperText-root-62"
              id="name-error-text"
            />
          </div>
          <div
            className="MuiCollapse-container-4"
            style={
              Object {
                "minHeight": "0px",
              }
            }
          >
            <div
              className="MuiCollapse-wrapper-6"
            >
              <div
                className="MuiCollapse-wrapperInner-7"
              >
                <p
                  className="MuiTypography-root-8 MuiTypography-body1-17 MuiTypography-colorError-31"
                />
              </div>
            </div>
          </div>
          <div
            className="CreateLunchspace-actionsContainer-2"
          >
            <button
              className="MuiButtonBase-root-88 MuiButton-root-66 MuiButton-contained-75 MuiButton-containedPrimary-76 MuiButton-raised-78 MuiButton-raisedPrimary-79 MuiButton-sizeLarge-86"
              disabled={false}
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="submit"
            >
              <span
                className="MuiButton-label-67"
              >
                Create Lunchspace
              </span>
              <span
                className="MuiTouchRipple-root-91"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="MuiCollapse-container-4"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper-6"
    >
      <div
        className="MuiCollapse-wrapperInner-7"
      >
        <span>
          Lunchspace 
        </span>
        <a
          href="//.lunchspace.de"
        >
          
        </a>
        <span>
           successful created.
        </span>
      </div>
    </div>
  </div>
</div>
`;
